### PR TITLE
Cross-signing [3/4] -- uploading signatures edition

### PR DESCRIPTION
**Pull request by uhoreg**
DO NOT MERGE until `uhoreg/e2e_cross-signing2` is merged, and the base of this PR changes to `develop`

Allows users to upload signatures using the new `/keys/signatures/upload` endpoint.

The `upload_signatures_for_device_keys` is really long.  I'm not sure what's the best way to make it more manageable.

### CODE REVIEW

File: synapse/handlers/e2e_keys.py [hunk 1]
```diff
@@ -511,6 +513,262 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+        self_device_ids = []
+
+        # split between checking signatures for own user and signatures for
+        # other users, since we verify them with different keys
+        if user_id in signatures:
+            self_signatures = signatures[user_id]
+            del signatures[user_id]
+            self_device_ids = list(self_signatures.keys())
+            try:
+                # get our self-signing key to verify the signatures
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: I haven't read the code in detail, but as a starting point for breaking up this function, can you not factor out the inside of this `try: except` block?

```python
try:
    signatures, failures = yield self._process_self_signatures(self_signatures))
    signature_list.extend(signatures)
    if failures:
        failures[user_id] = failures
except SynapseError as e:
    # etc
```

I'd also make `_process_self_signatures` start with:

```python
if not self_signatures:
    return [], None
```

so that you can remove the `if user_id in signatures` condition here.

(You can probably do something similar for the other_users case).

File: synapse/handlers/e2e_keys.py [hunk 2]
```diff
@@ -511,6 +513,262 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+                        signature = device["signatures"][user_id][self_signing_key_id]
+                        signature_list.append(
+                            (self_signing_key_id, user_id, device_id, signature)
+                        )
+                    except SynapseError as e:
+                        failures.setdefault(user_id, {})[
+                            device_id
+                        ] = _exception_to_failure(e)
+            except SynapseError as e:
+                failures[user_id] = {
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: kinda wondering under what circumstances we are likely to hit this case, and whether it is worth special handling rather than just failing the whole request.

Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: I'm not sure.  It's probably a spec discussion about whether the server can fail the whole request, and under what conditions.

**Comment by uhoreg** [comment]
There seems to be an intermittent sytest is failure.  https://github.com/matrix-org/synapse/pull/5726/commits/b6e3decd2b0c3608035f37bc6f9dcda9dd974322 failed with

    not ok 268 Real non-joined user doesn't get events before room made world_readable
    # Started: 2019-09-07 17:27:56.835
    # Ended: 2019-09-07 17:28:06.949
    # Timed out waiting for test at ./run-tests.pl line 727.
    # 0.091850: Registered new user @anon-20190907_172406-230:localhost:8828
    # 0.108166: Registered new user @anon-20190907_172406-231:localhost:8828
    # 0.781076: room_id=!dtQHGsYefGEhtkVyUv:localhost:8828

but https://github.com/matrix-org/synapse/pull/5726/commits/d3f2fbcfe577f42d0208d15a57bd66e56186742a (which only adds documentation) passed.  I suspect that it is unrelated to my changes.

I think this is ready for review again.

**Comment by richvdh** [comment]
> There seems to be an intermittent sytest is failure. b6e3dec failed with

Indeed. I've raised https://github.com/matrix-org/sytest/issues/696.

**Comment by richvdh** [comment]
please could you link to the MSC or spec this is implementing?

Also: can this target develop rather than an intermediate branch?

**Comment by richvdh** [comment]
> Also: can this target develop rather than an intermediate branch?

I've realised that the earlier MSCs have landed on said intermediate branch. I'm not a fan of long-lived feature branches: can we get it merged to develop?

**Comment by uhoreg** [comment]
> please could you link to the MSC or spec this is implementing?

https://github.com/matrix-org/matrix-doc/pull/1756

> I've realised that the earlier PRs have landed on said intermediate branch. I'm not a fan of long-lived feature branches: can we get it merged to develop?

@erikjohnston had suggested that I merge to a separate branch rather than develop.  I can do either way, so the two of you can ~fight it out~ discuss it, and let me know what I should do.

**Comment by richvdh** [comment]
> > please could you link to the MSC or spec this is implementing?
> 
> [matrix-org/matrix-doc#1756](https://github.com/matrix-org/matrix-doc/pull/1756)

thanks.
 
> > I've realised that the earlier PRs have landed on said intermediate branch. I'm not a fan of long-lived feature branches: can we get it merged to develop?
> 
> @erikjohnston had suggested that I merge to a separate branch rather than develop. I can do either way, so the two of you can ~fight it out~ discuss it, and let me know what I should do.

I've [discussed](https://matrix.to/#/!yZHTGeDKZUeKaqeTeU:matrix.org/$15681128341311URyZV:sw1v.org?via=matrix.org&via=t2bot.io&via=vector.modular.im) this with erik a bit, and I think our position goes like this: we prefer to avoid feature branches (especially long-lived ones) because:
 * if things are in develop, it saves having to repeatedly merge develop into the branch (or run the risk of it bit-rotting)
 * having a feature branch makes it easy to walk into a trap where PRs can be of poor quality, or introduce bugs, etc, because it's "only a feature branch", which means you rapidly lose track of what is left to be done to clean it up. (Not an issue in this case, I hasten to point out: just a natural consequence.)
 * in general, it's preferable to think about how to break a large change up so that it can safely land incrementally, rather than (say) splitting it up by file where none of it can land until all of it does - which again makes it easy to lose track of what's left to be done (and in many ways isn't much of an improvement over a single huge PR, since you still end up with the cognitive load of having to hold the whole change in your head at once). Again a feature branch makes it easier to fall into this trap.

On the other hand:
 * if it's really impossible to land a large change incrementally, we certainly prefer a feature branch with a few PRs over a 1000-line diff (I'm unconvinced that this is a real thing, though).
 * If you're making DB changes, and there's likely to be more change down the line, that's an argument for keeping things on a feature branch, since if you discover that the original DB schema was a bit wrong, you have to write a migration if the first PR has landed on develop.

TL;DR: If there are no particular reasons why the changes so far can't land on develop, I'd like to see them merged. If there *are* such reasons, could you identify them?

File: synapse/handlers/e2e_keys.py [hunk 3]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+
+        return signature_list, failures
+
+    @defer.inlineCallbacks
+    def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None):
+        key = yield self.store.get_e2e_cross_signing_key(
+            user_id, key_type, from_user_id
+        )
+        if key is None:
+            logger.error("no %s key found for %s", key_type, user_id)
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: don't think this is really an error (which indicates something has exploded within synapse). Possibly info, but really I think it's debug?

(likewise a bunch of other places that you log at error)

File: synapse/handlers/e2e_keys.py [hunk 4]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+            reasons
+        """
+        signature_list = []
+        failures = {}
+        if not signatures:
+            return signature_list, failures
+
+        try:
+            # get our self-signing key to verify the signatures
+            self_signing_key, self_signing_key_id, self_signing_verify_key = yield self._get_e2e_cross_signing_verify_key(
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: self_signing_key appears to be unused. I suggest making this clear with:

```suggestion
            _, self_signing_key_id, self_signing_verify_key = yield self._get_e2e_cross_signing_verify_key(
```

File: synapse/handlers/e2e_keys.py [hunk 5]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+                user_id, "self_signing"
+            )
+
+            # get our master key, since it may be signed
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: what does "it" refer to ?

File: synapse/handlers/e2e_keys.py [hunk 6]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+            # was sent if the device was signed
+            devices = yield self.store.get_e2e_device_keys([(user_id, None)])
+
+            if user_id not in devices:
+                raise SynapseError(404, "No device keys found", Codes.NOT_FOUND)
+
+            devices = devices[user_id]
+        except SynapseError as e:
+            failures[user_id] = {
+                device: _exception_to_failure(e) for device in signatures.keys()
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: can you do `_exception_to_failure` once rather than for each device?

File: synapse/handlers/e2e_keys.py [hunk 7]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+        for device_id, device in signatures.items():
+            try:
+                if "signatures" not in device or user_id not in device["signatures"]:
+                    # no signature was sent
+                    raise SynapseError(
+                        400, "Invalid signature", Codes.INVALID_SIGNATURE
+                    )
+
+                if device_id == master_verify_key.version:
+                    # we have master key signed by devices: for each
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: can you try to factor something out to another function here? 7 levels of indentation are too many.

File: synapse/handlers/e2e_keys.py [hunk 8]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+
+        for device_id, device in signatures.items():
+            try:
+                if "signatures" not in device or user_id not in device["signatures"]:
+                    # no signature was sent
+                    raise SynapseError(
+                        400, "Invalid signature", Codes.INVALID_SIGNATURE
+                    )
+
+                if device_id == master_verify_key.version:
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: seems like we should pull out

```
# signatures being uploaded by this user for this device
self_sigs = device["signatures"][user_id]
```

here, and get rid of some duplication below.

File: synapse/handlers/e2e_keys.py [hunk 9]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+                    # of the signatures on the master key succeed, or
+                    # all fail.  So loop over the signatures and add
+                    # them to a separate signature list.  If everything
+                    # works out, then add them all to the main
+                    # signature list.  (In practice, we're likely to
+                    # only have only one signature anyways.)
+                    master_key_signature_list = []
+                    sigs = device["signatures"]
+                    for signing_key_id, signature in sigs[user_id].items():
+                        alg, signing_device_id = signing_key_id.split(":", 1)
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: alg is unused:
```suggestion
                        _, signing_device_id = signing_key_id.split(":", 1)
```

File: synapse/handlers/e2e_keys.py [hunk 10]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+                        if (
+                            signing_device_id not in devices
+                            or signing_key_id
+                            not in devices[signing_device_id]["keys"]["keys"]
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: ugh, `["keys"]["keys"]` is horrible, as is the `stored_device = devices[device_id]["keys"]` below. Any chance I can get you to take on a preliminary refactor of `get_e2e_device_keys` to return something slightly saner?

File: synapse/handlers/e2e_keys.py [hunk 11]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+
+                        # get the key and check the signature
+                        pubkey = devices[signing_device_id]["keys"]["keys"][
+                            signing_key_id
+                        ]
+                        verify_key = decode_verify_key_bytes(
+                            signing_key_id, decode_base64(pubkey)
+                        )
+                        _check_device_signature(user_id, verify_key, device, master_key)
+                        device["signatures"] = sigs
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: what's going on here? I thought `sigs` was already the same as `device["signatures"]` ?

File: synapse/handlers/e2e_keys.py [hunk 12]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+        except SynapseError as e:
+            failure = _exception_to_failure(e)
+            for user, devicemap in signatures.items():
+                failures[user] = {device_id: failure for device_id in devicemap.keys()}
+            return signature_list, failures
+
+        for user, devicemap in signatures.items():
+            device_id = None
+            try:
+                # get the user's master key, to make sure it matches
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: > to make sure it matches what was sent

don't we rather need to get it to check the signature?

File: synapse/handlers/e2e_keys.py [hunk 13]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+                # what was sent
+                stored_key, stored_key_id, _ = yield self._get_e2e_cross_signing_verify_key(
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: call these `master_key`, `master_key_id` ?

File: synapse/handlers/e2e_keys.py [hunk 14]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+            user_signing_key, user_signing_key_id, user_signing_verify_key = yield self._get_e2e_cross_signing_verify_key(
+                user_id, "user_signing"
+            )
+        except SynapseError as e:
+            failure = _exception_to_failure(e)
+            for user, devicemap in signatures.items():
+                failures[user] = {device_id: failure for device_id in devicemap.keys()}
+            return signature_list, failures
+
+        for user, devicemap in signatures.items():
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: there are two many users flying around in this method. Can we distinguish between the uploading user and the target user?

File: synapse/handlers/e2e_keys.py [hunk 15]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+                # get the user's master key, to make sure it matches
+                # what was sent
+                stored_key, stored_key_id, _ = yield self._get_e2e_cross_signing_verify_key(
+                    user, "master", user_id
+                )
+
+                # make sure that the user's master key is the one that
+                # was signed (and no others)
+                device_id = stored_key_id.split(":", 1)[1]
+                if device_id not in devicemap:
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: Do we really need to accept the case when extra devices are signed? I feel like this might be more easily expressed as:

```
if devicemap.keys() != {device_id}:
    raise NotFoundError(...)
```

File: synapse/handlers/e2e_keys.py [hunk 16]
```diff
@@ -636,7 +911,58 @@ def _check_cross_signing_key(key, user_id, key_type, signing_key=None):
...
 
 
+def _check_device_signature(user_id, verify_key, signed_device, stored_device):
+    """Check that a device signature is correct and matches the copy of the device
+    that we have.  Throws an exception if an error is detected.
+
+    Args:
+        user_id (str): the user ID whose signature is being checked
+        verify_key (VerifyKey): the key to verify the device with
+        signed_device (dict): the signed device data
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: it would be useful to document (some of) the expected keys for this and stored_device. Do they match an object in the spec? if so, which?

File: synapse/handlers/e2e_keys.py [hunk 17]
```diff
@@ -636,7 +911,58 @@ def _check_cross_signing_key(key, user_id, key_type, signing_key=None):
...
+
+    signature = signed_device["signatures"][user_id][key_id]
+
+    # make sure that the device submitted matches what we have stored
+    del signed_device["signatures"]
+    # use pop to avoid exception if key doesn't exist
+    signed_device.pop("unsigned", None)
+    stored_device.pop("signatures", None)
+    stored_device.pop("unsigned", None)
+    if signed_device != stored_device:
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: I'm not sure this does what you want it to do, though tbh I'm not sure what you want it to do. 

comparisons are shallow in python...

File: synapse/handlers/e2e_keys.py [hunk 18]
```diff
@@ -636,7 +911,58 @@ def _check_cross_signing_key(key, user_id, key_type, signing_key=None):
...
+        or user_id not in signed_device["signatures"]
+        or key_id not in signed_device["signatures"][user_id]
+    ):
+        logger.error("upload signature: user not found in signatures")
+        raise SynapseError(400, "Invalid signature", Codes.INVALID_SIGNATURE)
+
+    signature = signed_device["signatures"][user_id][key_id]
+
+    # make sure that the device submitted matches what we have stored
+    del signed_device["signatures"]
```
Comment (richvdh) [inline synapse/handlers/e2e_keys.py]: if we're going to modify the inputs, that needs calling out in big letters on the docstring.

File: synapse/handlers/e2e_keys.py [hunk 19]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
...
+
+        Args:
+            user_id (string): the user uploading the signatures
+            signatures (dict[string, dict[string, dict]]): map of users to
+                devices to signed keys
+        """
+        failures = {}
+
+        # signatures to be stored.  Each item will be a tuple of
+        # (signing_key_id, target_user_id, target_device_id, signature)
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/e2e_keys.py]: actually let me put it more strongly: I suggest using `attrs` for this thing.

[the remark this answers is not in the document]
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: It must be good.  It has a testimonial from one "Amber Hawkie Brown". ;)

File: synapse/handlers/e2e_keys.py (hunk 5 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: "it" refers to the master key

File: synapse/handlers/e2e_keys.py (hunk 11 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: It's because `_check_device_signature` modifies `device`.  But I'm changing it so that it doesn't.

File: synapse/handlers/e2e_keys.py (hunk 12 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: No, in this function, we checking signatures on other users' master keys, made by our user-signing key.  (I'm adding a note in the docstring.)

File: synapse/handlers/e2e_keys.py (hunk 17 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: It seems to be deep

```python
>>> x = {"a": {"b": {"c": "d"}}}
>>> x == {"a": {"b": {"c": "d"}}}
True
>>> x == {"a": {"b": {"c": "e"}}}
False
```

File: synapse/handlers/e2e_keys.py (hunk 16 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: They're device or cross-signing key objects.  I've added some words to the doc string, so hopefully it's a bit clearer?

File: synapse/handlers/e2e_keys.py [hunk 20]
```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
 
         return {}
 
+    @defer.inlineCallbacks
+    def upload_signatures_for_device_keys(self, user_id, signatures):
+        """Upload device signatures for cross-signing
+
+        Args:
+            user_id (string): the user uploading the signatures
+            signatures (dict[string, dict[string, dict]]): map of users to
```
[the remark this answers is not in the document]
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: Added comment, and added some validation.

File: synapse/handlers/e2e_keys.py (hunk 8 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: I'm not sure what you're asking for here.

File: synapse/handlers/e2e_keys.py (hunk 10 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: I can do so, if you have suggestions for how to fix it.  It looks like `get_e2e_device_keys` is only used here and in `query_local_devices`.  There, it needs `device_info["keys"]` and `device_info["device_display_name"]` and modifies `device_info["keys"]`to add the display name.  Should I just make `query_local_devices` in change of adding in the device display name, and then have `devices[signing_device_id]["keys"]` become just `devices[signing_device_id]`?

File: synapse/handlers/e2e_keys.py (hunk 15 above)
Comment (uhoreg) [reply synapse/handlers/e2e_keys.py]: It seems to me that we should accept the data that we can rather than failing the whole thing.

**Comment by uhoreg** [comment]
Sytest failures seem to be new 3pid-related stuff, which will hopefully get fixed when I merge develop in (again).  The synapse unit test failure seems to be my old friend `test_replace_master_key` which is the unit test failure from https://github.com/matrix-org/synapse/pull/5769 that I had thought that I had fixed by using stream ID generators.

Aside from that, though, I think this is ready for review again now.  I can fix https://github.com/matrix-org/synapse/pull/5726#discussion_r327798471 if what I've written there makes sense.

File: synapse/handlers/e2e_keys.py (hunk 10 above)
Comment (richvdh) [reply synapse/handlers/e2e_keys.py]: that sounds sensible, I think. I'd love it if it could be a preliminary PR rather than rolled into this one, though.

File: synapse/handlers/e2e_keys.py (hunk 15 above)
Comment (richvdh) [reply synapse/handlers/e2e_keys.py]: I've lost track of what it is we're processing here and how it relates to the API, so I'll take your word for it.

File: synapse/handlers/e2e_keys.py (hunk 17 above)
Comment (richvdh) [reply synapse/handlers/e2e_keys.py]: you're right. I'm a crank.

**Comment by uhoreg** [comment]
> As well as the nits above, and fixing whatever's wrong with the CI: can we target develop as per #5726 (comment)?

Yes, I'll do that, although I want to try to fix the CI before I merge the previous PRs into develop, so that we don't have a broken CI in develop.

**Comment by uhoreg** [comment]
I think I've addressed all the issues, and I've re-merged develop so the tests pass.  I've also changed this PR so that it targets develop, now that the previous PRs have been merged in.
